File: frontend/src/types/report.ts
export type ReportType = 'secr' | 'cbam_declaration' | 'executive_summary';
export type ReportStatus = 'draft' | 'generated' | 'submitted' | 'approved';

export interface Report {
  id: string;
  user_id: string;
  title: string;
  type: ReportType;
  period: string;
  status: ReportStatus;
  total_emissions_tco2e: number | null;
  total_cbam_charge_eur: number | null;
  notes: string | null;
  start_date: string | null;
  end_date: string | null;
  created_date: string;
}

export interface ReportCreatePayload {
  title: string;
  type: ReportType;
  period: string;
  status?: ReportStatus;
  total_emissions_tco2e?: number;
  total_cbam_charge_eur?: number;
  notes?: string;
  start_date?: string;
  end_date?: string;
}

export interface ReportGeneratePayload {
  title: string;
  type: ReportType;
  period: string;
  start_date?: string;
  end_date?: string;
}

// --- Aggregation types ---

export interface ScopeBreakdown {
  scope: string;
  label: string;
  emissions_tco2e: number;
  description: string;
}

export interface CategoryBreakdown {
  category: string;
  scope: string;
  emissions_tco2e: number;
  share_pct: number;
}

export interface ActivityDetail {
  activity_name: string;
  scope: string;
  source: string | null;
  quantity: number;
  unit: string;
  co2e_tco2e: number;
  activity_date: string | null;
}

export interface CBAMCategoryBreakdown {
  category: string;
  total_qty_tonnes: number;
  embedded_emissions_tco2e: number;
  cbam_charge_eur: number;
}

export interface CBAMImportDetail {
  product_name: string;
  hscn_code: string;
  origin_country: string;
  supplier_name: string | null;
  quantity_tonnes: number;
  embedded_emissions: number;
  cbam_charge_eur: number;
  declaration_status: string;
}

export interface OrganizationInfo {
  name: string;
  industry: string | null;
  country: string | null;
  reporting_year: number | null;
  base_year: number | null;
  reduction_target_pct: number | null;
}

export interface ReportAggregation {
  organization: OrganizationInfo | null;
  total_emissions_tco2e: number;
  scope_breakdown: ScopeBreakdown[];
  category_breakdown: CategoryBreakdown[];
  activities: ActivityDetail[];
  activity_count: number;
  total_cbam_charge_eur: number;
  total_embedded_emissions_tco2e: number;
  cbam_category_breakdown: CBAMCategoryBreakdown[];
  cbam_imports: CBAMImportDetail[];
  cbam_import_count: number;
  pending_declarations: number;
}

export interface PaginatedReportsResponse {
  items: Report[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

File: backend/app/schemas/report.py
import uuid
from datetime import date, datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel


class ReportCreate(BaseModel):
    title: str
    type: str
    period: str
    status: Optional[str] = "draft"
    total_emissions_tco2e: Optional[float] = None
    total_cbam_charge_eur: Optional[float] = None
    notes: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class ReportUpdate(BaseModel):
    title: Optional[str] = None
    type: Optional[str] = None
    period: Optional[str] = None
    status: Optional[str] = None
    total_emissions_tco2e: Optional[float] = None
    total_cbam_charge_eur: Optional[float] = None
    notes: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class ReportResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    type: str
    period: str
    status: str = "draft"
    total_emissions_tco2e: Optional[float] = None
    total_cbam_charge_eur: Optional[float] = None
    notes: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    created_date: datetime

    class Config:
        from_attributes = True


class ReportGenerateRequest(BaseModel):
    title: str
    type: str
    period: str
    start_date: Optional[date] = None
    end_date: Optional[date] = None




class ScopeBreakdown(BaseModel):
    scope: str
    label: str
    emissions_tco2e: float
    description: str


class CategoryBreakdown(BaseModel):
    category: str
    scope: str
    emissions_tco2e: float
    share_pct: float


class ActivityDetail(BaseModel):
    activity_name: str
    scope: str
    source: Optional[str] = None
    quantity: float
    unit: str
    co2e_tco2e: float
    activity_date: Optional[str] = None


class CBAMCategoryBreakdown(BaseModel):
    category: str
    total_qty_tonnes: float
    embedded_emissions_tco2e: float
    cbam_charge_eur: float


class CBAMImportDetail(BaseModel):
    product_name: str
    hscn_code: str
    origin_country: str
    supplier_name: Optional[str] = None
    quantity_tonnes: float
    embedded_emissions: float
    cbam_charge_eur: float
    declaration_status: str


class OrganizationInfo(BaseModel):
    name: str
    industry: Optional[str] = None
    country: Optional[str] = None
    reporting_year: Optional[int] = None
    base_year: Optional[int] = None
    reduction_target_pct: Optional[float] = None


class ReportAggregationResponse(BaseModel):
    """Full aggregation data returned by /preview and /{id}/data endpoints."""
    organization: Optional[OrganizationInfo] = None

    
    total_emissions_tco2e: float = 0.0
    scope_breakdown: List[ScopeBreakdown] = []
    category_breakdown: List[CategoryBreakdown] = []
    activities: List[ActivityDetail] = []
    activity_count: int = 0

    
    total_cbam_charge_eur: float = 0.0
    total_embedded_emissions_tco2e: float = 0.0
    cbam_category_breakdown: List[CBAMCategoryBreakdown] = []
    cbam_imports: List[CBAMImportDetail] = []
    cbam_import_count: int = 0
    pending_declarations: int = 0
